design-proposals/dynamic-primary-pod-nic-name.md

```diff
@@ -0,0 +1,298 @@
+The default primary pod interface is often named `eth0`.
+However, users might require custom network setups where the pod’s primary interface is not named `eth0`.
+Currently, KubeVirt attaches the VM’s interface destined to the pod network to the pod’s `eth0` interface.
+
+When creating a VM with secondary networks, KubeVirt specifies the pod interfaces’ names for all secondary interfaces.
+It is currently not possible to specify the primary interface’s name when templating the virt-launcher pod, since there
+is no K8s API that enables it.
+
+## Goals
+
+- Allow the primary pod interface name to be dynamically determined based on the environment.
+- Ensure backward compatibility, so existing VMs with `eth0` as the primary interface do not break.
+- Handle the scenario of a non-eth0 primary interface with a binding plugin.
+
+## Non Goals
+
+- Change Kubernetes' API.
+- Change Multus' API.
+
+## Definition of Users
+
+- Cluster Administrator
+- VM owner
+
+## User Stories
+
+- As a cluster administrator I want to use a CNI that configures pods’ primary interface to a name other than `eth0`.
+
+- As a VM owner I want my VMs to be connected to the cluster's default network, as they do today.
+
+- As a VM owner I want a correct interface status reporting.
+
+## Repos
+
+- https://github.com/kubevirt/kubevirt
+
+# Design
+
+## Using Multus’ network-status annotation
+
+When KubeVirt is used in a cluster that has Multus as its default CNI, Multus adds the
+`k8s.v1.cni.cncf.io/network-status` annotation
+to the virt-launcher pod following its network attachment process.
+
+This annotation includes information about all the network interfaces attached to the pod, for example:
+
+```json
+[
+  {
+    "name": "k8s-pod-network",
+    "ips": [
+      "10.244.196.146",
+      "fd10:244::c491"
+    ],
+    "default": true,
+    "dns": {}
+  },
+  {
+    "name": "meganet",
+    "interface": "pod7e0055a6880",
+    "mac": "8a:37:d9:e7:0f:18",
+    "dns": {}
+  }
+]
+```
+
+This annotation could be parsed to infer the default (primary) network interface instead of relying on a hard coded
+interface name (like eth0).
+
+If the network-status annotation is missing or doesn’t specify a default interface, KubeVirt will fall back to using
+`eth0` as the default interface, maintaining backward compatibility.
+
+> [!NOTE]
+> The VMI controller currently uses this annotation for:
+> - Updating the VMI status interfaces section with Multus info-source.
+> - Creating DownwardAPI for SR-IOV NICs and certain network binding plugins.
+
+
+Since this information is also required in virt-handler and virt-launcher, it could be passed on the
+VirtualMachineInstance object.
+
+> [!NOTE]
+> virt-handler and virt-launcher do not have access to the Pod object.
+
+## Alternative 1 - Report Pod Interface Name Per Interface
+
+Currently, the `VirtualMachineInstanceStatus` struct holds a slice of interface status objects:
```

Comment:
> Using the k8s subresource to define something is not a good idea, since depending on how it's implemented it may be ignored by reconcile.
> 

AFAIK, ATM the VirtualMachineInstance CRD does not have a `/status` sub-resource [1].
Could you please elaborate on the scenarios on which updating the VMI status could be ignored?

> Also do we really need to configure this PR vm or per network ? I doubt a per VM make much sense.

Currently both per-interface and once per VMI options are described.
Both has their pros and cons.

Could you please elaborate why do you think having the information per VM does not make sense?

[1] https://github.com/kubevirt/kubevirt/pull/3889

Reply:
> > > Using the k8s subresource to define something is not a good idea, since depending on how it's implemented it may be ignored by reconcile.
> 
> I do not understand this. The status of any resource is always reconciled, that is the idea of having a status that represents the current state. If it is a subresource or not is an implementation detail that is suppose to improve safety.
> 

If you have a CRD with a status the put a condition about "success" and a Reconcile cycle that update that, you don't want to enter that reconcile if you update it since it will be a loop, so depending on the implementation that is skip, but maybe not the case for kubevirt.

> > AFAIK, ATM the VirtualMachineInstance CRD does not have a `/status` sub-resource [1].
> 
> The VMI has a status subresource.

Reply:
> > > Using the k8s subresource to define something is not a good idea, since depending on how it's implemented it may be ignored by reconcile.
> 
> I do not understand this. The status of any resource is always reconciled, that is the idea of having a status that represents the current state. If it is a subresource or not is an implementation detail that is suppose to improve safety.
> 
> > AFAIK, ATM the VirtualMachineInstance CRD does not have a `/status` sub-resource [1].
> 
> The VMI has a status subresource.

The VMI CRD does not have a status sub-resource.
Please see:

https://github.com/kubevirt/kubevirt/blob/dc6b76f494ffdb95305fed5ce038963539d469a1/pkg/virt-operator/resource/generate/components/crds.go#L167

in comparison to the VirtualMachine CRD that does have it:
https://github.com/kubevirt/kubevirt/blob/dc6b76f494ffdb95305fed5ce038963539d469a1/pkg/virt-operator/resource/generate/components/crds.go#L205

Reply:
> Currently both per-interface and once per VMI options are described. Both has their pros and cons.
> 
> Could you please elaborate why do you think having the information per VM does not make sense?
> 
> [1] [kubevirt/kubevirt#3889](https://github.com/kubevirt/kubevirt/pull/3889)

As far as I have see primary interface is common between all the VMs on the same network, so it do not make sense to configure it per VM.

Reply:
> The VMI CRD does not have a status sub-resource.

This is surprising.

> As far as I have see primary interface is common between all the VMs on the same network, so it do not make sense to configure it per VM.

I am not familiar with an entity that is describing a VM collection per namespace.
The Kubevirt CR is at the cluster level and creating a new resource kind for this does not make sense.

But I also do not understand how is this even relevant for the line this comment thread refers to.
Are you suggesting another alternative? Can you suggest to add one at a proper location with all the details?